dialog/dialog/query-engine.md

```diff
@@ -166,25 +164,27 @@ Note that while this rule depends negatively on itself, it's able to do so safel
 
 As Dialog is intended for use in distributed and decentralized deployments, it is important ensure the use of collision resistant identifiers when referring to tuples. For this purpose, a content addressing scheme is leveraged, wherein facts are associated with a content ID (CID) computed from their structure. The details behind this computation are available in [serialization](./serialization.md).
 
+(Note: this may not actually be the right way of thinking about this, because in many cases, these CIDs are for tuples derived from the EVAC tuples in IPFS, rather than the EVAC tuples themselves. Instead we might also want to accumulate the provenance of variables across joins, and expose a predicate that queries over that provenance. This provenance tracking might even happen conditionally, based on whether it's needed by the program)
```

Comment:
Do I understand this correctly:
If the provenance tracking is optimized per-program, wouldn't that mean that the CID you use to refer to some ... fact/thing depends on the program you're using to interpret a set of facts?
That would mean we're not in a world where everyone can choose their own program for interacting with the tuple store. People would have to agree on the programs.

Reply:
This is partly correct, and an excellent point to bring up! I address some of this in here: https://discord.com/channels/478735028319158273/1033502043656171561/1035339517021921280

This point is why I think we need to distinguish between the CID for an arbitrary tuple, and the CIDs of the source EVAC tuples in a tuple's provenance: in the first case, we're dealing with relations with a program-defined schema, whereas in the second, it's always going to be small granular EVAC tuples over a fixed schema.

There is still some open discussion around whether we want to make use of any dataflow analysis or anything or constrain such entanglement in useful ways, and I think Edelweiss is a good example to draw on here: for example, perhaps it only makes sense to allow entangling against a non-source CID as part of intermediate transient relations. Then they'd be guaranteed to never leave the local device, and there's no risk of writing those pointers out to the network.

Part of the plan is for us (or anyone!) to expose a small library of rules for implementing CRDTs over these EVAC tuples, which essentially exposes an interface over the DAG for interpreting it. It's entirely possible that a malformed or malicious program might write tuples which conform to that interface in unexpected or undesirable ways, but the hypothesis is that our use of a merkle DAG here gives us resilience in the face of such tampering: either they'll insert meaningless data which will be ignored, or they'll insert meaningful garbage, which is indistinguishable from a valid use of the interface anyways.

Does that make sense?